crates/red_knot_test/src/lib.rs

```diff
@@ -121,6 +140,42 @@ fn run_test(db: &mut db::Db, test: &parser::MarkdownTest) -> Result<(), Failures
         })
         .collect();
 
+    // Create a custom typeshed `VERSIONS` file if none was provided.
+    if let Some(ref typeshed_path) = custom_typeshed_path {
+        if !has_custom_versions_file {
+            let versions_file = typeshed_path.join("stdlib/VERSIONS");
+            let contents = typeshed_files
+                .iter()
+                .fold(String::new(), |mut content, path| {
+                    // This is intentionally kept simple:
+                    let module_path = path
```

Comment:
nit: strictly speaking we're constructing a module _name_ from the module path

```suggestion
                    let module_name = path
```

The logic here LGTM though, assuming all paths in `typeshed_files` are already relative to `/typeshed/stdlib`.